nexus/reconfigurator/execution/src/datasets.rs

```diff
@@ -133,121 +133,91 @@ pub(crate) async fn ensure_dataset_records_exist(
-                num_unchanged += 1;
-                continue;
+
+    // Filter down to in-service Crucible datasets.
+    let wanted_datasets = bp_datasets.filter_map(|d| {
+        match d.disposition {
+            BlueprintDatasetDisposition::InService => (),
+            BlueprintDatasetDisposition::Expunged => return None,
+        }
+        // We only insert records for Crucible datasets.
+        let dataset = match (&d.kind, d.address) {
+            (DatasetKind::Crucible, Some(addr)) => CrucibleDataset::new(
+                d.id,
+                d.pool.id().into_untyped_uuid(),
+                addr,
+            ),
+            (DatasetKind::Crucible, None) => {
+                // This should be impossible! Ideally we'd prevent it
+                // statically, but for now just fail at runtime.
+                return Some(Err(anyhow!(
+                    "invalid blueprint dataset: \
+                     {} has kind Crucible but no address",
+                    d.id
+                )));
             }
-            num_updated += 1;
-            "update"
-        } else {
-            num_inserted += 1;
-            "insert"
+            _ => return None,
         };
+        Some(Ok(dataset))
+    });
+
+    for dataset in wanted_datasets {
+        let dataset = dataset?;
+        let id = dataset.id();
+
+        // If this dataset already exists, do nothing.
+        if existing_datasets.contains_key(&id) {
+            num_unchanged += 1;
+            continue;
+        }
 
-        let dataset = Dataset::from(bp_dataset.clone());
         datastore
-            .dataset_upsert_if_blueprint_is_current_target(
+            .crucible_dataset_upsert_if_blueprint_is_current_target(
                 &opctx, bp_id, dataset,
             )
             .await
             .with_context(|| {
                 format!("failed to upsert dataset record for dataset {id}")
             })?;
+        num_inserted += 1;
 
         info!(
             opctx.log,
-            "ensuring dataset record in database";
-            "action" => action,
+            "ensuring Crucible dataset record in database";
+            "action" => "insert",
             "id" => %id,
-            "kind" => ?kind,
         );
     }
 
-    for bp_dataset in unwanted_datasets {
-        if existing_datasets.remove(&bp_dataset.id).is_some() {
-            if matches!(
-                bp_dataset.kind,
-                omicron_common::disk::DatasetKind::Crucible
-            ) {
-                // Region and snapshot replacement cannot happen without the
-                // database record, even if the dataset has been expunged.
-                //
-                // This record will still be deleted, but it will happen as a
-                // part of the "decommissioned_disk_cleaner" background task.
-                continue;
-            }
-
-            datastore
-                .dataset_delete_if_blueprint_is_current_target(
```

Comment:
Rad to see this unified, it felt gross as it was written (deletion of a table row happening in multiple possible spots). Thanks again for this change.